src/compiler/transformers/esnext.ts

```diff
@@ -168,48 +162,58 @@ namespace ts {
             }
 
             const target = createConditional(
-                createLogicalOr(
-                    createStrictEquality(createAssignment(temp, leftExpression), createNull()),
-                    createStrictEquality(temp, createVoidZero())
-                ),
+                createNotNullCondition(capturedLeft, /*negate*/ true),
                 createVoidZero(),
-                rightExpression
+                rightExpression,
             );
             return thisArg ? createSyntheticReferenceExpression(target, thisArg) : target;
         }
 
-        function createNotNullCondition(node: Expression) {
+        function createNotNullCondition(expression: CapturedExpression, negate?: boolean) {
             return createBinary(
                 createBinary(
-                    node,
-                    createToken(SyntaxKind.ExclamationEqualsEqualsToken),
+                    expression.expression,
+                    createToken(negate ? SyntaxKind.EqualsEqualsEqualsToken : SyntaxKind.ExclamationEqualsEqualsToken),
                     createNull()
                 ),
-                createToken(SyntaxKind.AmpersandAmpersandToken),
+                createToken(negate ? SyntaxKind.BarBarToken : SyntaxKind.AmpersandAmpersandToken),
                 createBinary(
-                    node,
-                    createToken(SyntaxKind.ExclamationEqualsEqualsToken),
+                    expression.variable,
+                    createToken(negate ? SyntaxKind.EqualsEqualsEqualsToken : SyntaxKind.ExclamationEqualsEqualsToken),
                     createVoidZero()
                 )
             );
         }
 
         function transformNullishCoalescingExpression(node: BinaryExpression) {
-            const expressions: Expression[] = [];
-            let left = visitNode(node.left, visitor, isExpression);
-            if (!isIdentifier(left)) {
-                const temp = createTempVariable(hoistVariableDeclaration);
-                expressions.push(createAssignment(temp, left));
-                left = temp;
+            const captured = maybeCaptureInTempVariable(visitNode(node.left, visitor, isExpression));
+            return createConditional(
+                createNotNullCondition(captured),
+                captured.variable,
+                visitNode(node.right, visitor, isExpression),
+            );
+        }
+
+        interface CapturedExpression {
+            expression: Expression;
+            variable: Expression;
+        }
+
+        function maybeCaptureInTempVariable(expression: Expression): CapturedExpression {

```

Comment:
As nice as it is to have a reusable function for this behavior, v8 doesn't optimize away destructured return values. When emitting a large codebase that might make heavy use of optional chaining, we could end up creating a lot of short-lived `{ expression, variable }` objects which can result in more GC time. 

Given how small `maybeCaptureInTempVariable` is, it would probably be better to manually inline it into each of its call sites. One approach might be to create a function that checks the condition instead:

```ts
function shouldCaptureInTempVariable(expression: Expression) {
  return isIdentifier(expression) || ...; // from L205
}
function getCaptureVariable(shouldCapture: boolean, expression: Expression) {
  return shouldCapture ? createTempVariable(...) : expression;
}
function getCaptureExpression(shouldCapture: boolean, variable: Expression, expression: Expression) {
  return shouldCapture ? createAssignment(...) : expression;
}
```

And then inline the previous function:
```ts
// ({ expression, variable: thisArg } = maybeCaptureInTempVariable(expression));
const shouldCapture = shouldCaptureInTempVariable(expression);
thisArg = getCaptureVariable(shouldCapture, expression);
expression = getCaptureExpression(shouldCapture, variable, expression);
```

The upside is that `getCaptureVariable` and `getCaptureExpression` are small enough that V8 will probably inline them anyways (and we could also opt to just do the inlining ourselves), and we avoid the possibility of excess GC scavenges.

Reply:
I extracted `shouldCaptureInTempVariable` for reuse. The other code was way too simple to make a function for it so I inlined it at the call sites.
To reduce duplication I unified the code paths of `PropertyAccessExpression` and `ElementAccessExpression`